pkg/apis/pipeline/v1beta1/task_validation_test.go

```diff
@@ -552,6 +564,75 @@ func TestTaskSpecValidateError(t *testing.T) {
 			Message: `"string" type does not match default value's type: "array"`,
 			Paths:   []string{"params.task.type", "params.task.default.type"},
 		},
+	}, {
+		name: "param mismatching default/type 3",
+		fields: fields{
+			Params: []v1beta1.ParamSpec{{
+				Name:        "task",
+				Type:        v1beta1.ParamTypeArray,
+				Description: "param",
+				Default: v1beta1.NewObject(map[string]string{
+					"key1": "var1",
+					"key2": "var2",
+				}),
+			}},
+			Steps: validSteps,
+		},
+		expectedError: apis.FieldError{
+			Message: `"array" type does not match default value's type: "object"`,
+			Paths:   []string{"params.task.type", "params.task.default.type"},
+		},
+	}, {
+		name: "param mismatching default/type 4",
+		fields: fields{
+			Params: []v1beta1.ParamSpec{{
+				Name:        "task",
+				Type:        v1beta1.ParamTypeObject,
+				Description: "param",
+				Properties:  map[string]v1beta1.PropertySpec{"key1": {}},
+				Default:     v1beta1.NewArrayOrString("var"),
+			}},
+			Steps: validSteps,
+		},
+		expectedError: apis.FieldError{
+			Message: `"object" type does not match default value's type: "string"`,
+			Paths:   []string{"params.task.type", "params.task.default.type"},
+		},
+	}, {
+		name: "the spec of object type parameter misses providing properties",
+		fields: fields{
+			Params: []v1beta1.ParamSpec{{
+				Name:        "task",
+				Type:        v1beta1.ParamTypeObject,
+				Description: "param",
+			}},
+			Steps: validSteps,
+		},
+		expectedError: apis.FieldError{
+			Message: "The parameter of object type misses the definition of `properties`",
+			Paths:   []string{"params.task.type", "params.task.default.type"},
+		},
+	}, {
+		name: "keys defined in properties are missed in default",
+		fields: fields{
+			Params: []v1beta1.ParamSpec{{
+				Name:        "myobjectParam",
+				Description: "param",
+				Properties: map[string]v1beta1.PropertySpec{
+					"key1": {},
+					"key2": {},
+				},
+				Default: v1beta1.NewObject(map[string]string{
+					"key1": "var1",
+					"key3": "var1",
+				}),
+			}},
+			Steps: validSteps,
+		},
+		expectedError: apis.FieldError{
+			Message: fmt.Sprintf("Required keys %s for the parameter \"%s\" are not provided in default.", []string{"key2"}, "myobjectParam"),
+			Paths:   []string{""},
+		},
```

Comment:
+1 I think this PR is mixed with multiple stuff, for example, default values, but miss a big chunk of other validations. Maybe scope out this PR to a specific module / feature and that would make the PR easier to follow

Reply:
> I think we need more tests where we don't rely on `Defaults` to set the values.

This [validation](https://github.com/tektoncd/pipeline/pull/4786/files/0ca9635e51663e4916a9732edf2b7affc1c3855b#diff-988b00fa4f188a96ff5901c0bbc9f65d4a8355417185d4fe0ae97d8c14e04417R363) is to only check if keys are missed iff the `default` section is provided ([code](https://github.com/tektoncd/pipeline/pull/4786/files/0ca9635e51663e4916a9732edf2b7affc1c3855b#diff-988b00fa4f188a96ff5901c0bbc9f65d4a8355417185d4fe0ae97d8c14e04417R351-R353)). 

With that said, it's not relying on `default` to set values for the object param. `pkg/reconciler/taskrun/validate_resources.go` will be responsible for check if neither taskrun or default provides a value for a param.

Reply:
> +1 I think this PR is mixed with multiple stuff, for example, default values, but miss a big chunk of other validations. Maybe scope out this PR to a specific module / feature and that would make the PR easier to follow

For validation, this pr only includes task&taskrun level validation. Pipeline & pipeline level validation will be in future PRs.

For task&taskrun level validation, this PR includes most of validations except 
1. validate against entire object usage
> When providing values for strings, Task and Pipeline authors can access individual attributes of an object param; they cannot access the object as whole (we could add support for this later). (mentioned [here]((https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#using-objects-in-variable-replacement)) in the tep)

2. restrict the value of type in `PropertySpec` down to `string` only. ([mentioned](https://github.com/tektoncd/pipeline/pull/4786#discussion_r863921383) by billy)

Considering the amount of changes in this pr and amount of work for 1&2, I am thinking to add 1&2 in two different PRs later.

Reply:
will try to make future prs more modularized :)

Reply:
> This validation is to check if keys are missed in default iff the default section is provided (code).

💯  Let's add a test to confirm the behavior and make sure that assumption doesn't break in the future without us knowing.

Perhaps also add tests for TaskRun validation with object params as well? Otherwise totally agree with breaking things up into multiple pull requests!

Reply:
> 💯 Let's add a test to confirm the behavior and make sure that assumption doesn't break in the future without us knowing.

Good test case! Just added `myobjWithoutDefault` in `TestTaskValidate` function, which is object type and default is not provided. This valid param test makes sure it doesn't rely on `default` to set values for the object param. 

> Perhaps also add tests for TaskRun validation with object params as well? Otherwise totally agree with breaking things up into multiple pull requests!

Good catch! After adding the alpha flag check, test coverage for `taskrun_validation.go` went down by 1.4% 🤣 . I just added new test to test duplicate params which are object type to bring it back 100%. I think this is the only test needed to add because `taskrun_validation.go` doesn't implement any validation against type alignment etc. Let me know otherwise.